lib/rqrcode_core/qrcode/qr_code.rb

```diff
@@ -152,69 +168,59 @@ class QRCodeRunTimeError < RuntimeError; end
   class QRCode
     attr_reader :modules, :module_count, :version
 
-    # Expects a string to be parsed in, other args are optional
+    # Expects a string or array (for multi-segment encoding) to be parsed in, other args are optional
     #
-    #   # string - the string you wish to encode
-    #   # size   - the size (Integer) of the qrcode (defaults to smallest size needed to encode the string)
+    #   # data - the string, QRSegment or array of QRSegments you wish to encode
+    #   # size   - the size (Integer) of the qrcode (defaults to smallest size needed to encode the data)
     #   # level  - the error correction level, can be:
     #      * Level :l 7%  of code can be restored
     #      * Level :m 15% of code can be restored
     #      * Level :q 25% of code can be restored
     #      * Level :h 30% of code can be restored (default :h)
-    #   # mode   - the mode of the qrcode (defaults to alphanumeric or byte_8bit, depending on the input data):
+    #   # mode   - the mode of the qrcode (defaults to alphanumeric or byte_8bit, depending on the input data, only used when data is a string):
     #      * :number
     #      * :alphanumeric
     #      * :byte_8bit
     #      * :kanji
+    #      * :multi
     #
     #   qr = RQRCodeCore::QRCode.new('hello world', size: 1, level: :m, mode: :alphanumeric)
-    #
-
-    def initialize(string, *args)
-      if !string.is_a? String
-        raise QRCodeArgumentError, "The passed data is #{string.class}, not String"
-      end
+    #   segment_qr = QRCodeCore::QRCode.new(QRSegment.new(data: 'foo', mode: :byte_8bit))
+    #   multi_qr = RQRCodeCore::QRCode.new([QRSegment.new(data: 'foo', mode: :byte_8bit), QRSegment.new(data: 'bar1', mode: :alphanumeric)])
 
+    def initialize(data, *args)
       options = extract_options!(args)
+
       level = (options[:level] || :h).to_sym
+      max_size = options[:max_size] || QRUtil.max_size
 
-      if !QRERRORCORRECTLEVEL.has_key?(level)
-        raise QRCodeArgumentError, "Unknown error correction level `#{level.inspect}`"
+      @data = case data
+      when String
+        QRSegment.new(data: data, mode: options[:mode])
+      when Array
+        raise QRCodeArgumentError, "Array must contain QRSegments" if data.find { |seg| !seg.is_a?(QRSegment) }
+        data
```

Comment:
I think on reflection that we should instantiate the `QRSegment`'s in the initializer. From a user point of view I think:

```
multi_qr = RQRCodeCore::QRCode.new([
  {data: 'foo', mode: :byte_8bit}, 
  {data: 'bar1', mode: :alphanumeric}
])
```

is less to write and care about than

```
multi_qr = RQRCodeCore::QRCode.new([
  RQRCodeCore::QRSegment.new(data: 'foo', mode: :byte_8bit), 
  RQRCodeCore::QRSegment.new(data: 'bar1', mode: :alphanumeric)
])
```

and from the `RQRCode` point of view (whilst multi mode is quite niche) I think not having to add any extra code in that gem is a good thing, especially as I'd like to move eventually to having the various renderers be light weight and handle a single rendering `rqrcode_png`, `rqrcode_svg` etc. I get a lot of requests for different renderings of a QR Code and at the moment I am pushing back as I don't want to turn `RQRCode` into an unmanageable monolith.

Reply:
I think that's a good idea, I think it gives us the best of best worlds. I'll make the changes to do it this way.